### PR TITLE
separable read vs write schemas

### PR DESCRIPTION
**Pull request by jgraettinger**
**Description:**

Allow collections to have a `writeSchema` and `readSchema` as an alternative to the current combined `schema`.

Remove the capability for derivation transforms to define an alternative `source` schema, as this behavior is made redundant by collection-level read schemas.

**Workflow steps:**

Optionally transition current collections to have separate read & write schemas.

Existing built specifications will continue to function as intended.

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/826)

### CODE REVIEW

File: crates/assemble/src/npm/generators.rs [hunk 1]
```diff
@@ -16,10 +16,11 @@ pub fn module_types(
 ) -> String {
     let mut w = String::with_capacity(4096);
 
-    let doc_mapper = build_mapper(compiled, imports, &collection.schema, true);
+    let w_mapper = build_mapper(compiled, imports, &collection.write_schema, false);
+    let r_mapper = build_mapper(compiled, imports, &collection.read_schema, true);
```
Comment (willdonnelly) [inline crates/assemble/src/npm/generators.rs:20]: So the logic in this function generates types `SourceDocument` and `OutputDocument` but then only the `SourceDocuments` are re-exported as `<Transform>Source` later on, and the `OutputDocument` types are referenced directly in the actual derivation code. Do I have that right?

Comment (jgraettinger) [reply crates/assemble/src/npm/generators.rs:20]: Yep, that's right.

And in the combined-schema case we generate `Document` which aliases to `SourceDocument` & `OutputDocument`, which has the convenient property of not breaking existing derivations which import that `Document` type.

File: crates/schemalate/src/firebolt/firebolt_schema_builder.rs [hunk 2]
```diff
@@ -57,7 +57,7 @@ fn build_shape_from_schema(schema_str: &str) -> Result<Shape, Error> {
 pub fn build_firebolt_schema(binding: &Binding) -> Result<TableSchema, Error> {
     let fs = binding.field_selection.as_ref().unwrap();
     let projections = &binding.collection.as_ref().unwrap().projections;
-    let schema_str = &binding.collection.as_ref().unwrap().schema_json;
+    let schema_str = &binding.collection.as_ref().unwrap().write_schema_json;
```
Comment (willdonnelly) [inline crates/schemalate/src/firebolt/firebolt_schema_builder.rs:60]: Is the write schema the correct thing to be using here? If I understand correctly it seems like the possibly-more-restrictive read schema of the collection is what we'd want to reflect in a materialization.

**Review by willdonnelly — APPROVED** [review]
LGTM

File: crates/flowctl/src/draft/develop.rs [hunk 3]
```diff
@@ -99,11 +99,27 @@ pub async fn do_develop(
...
-                        &catalog_url.join(&format!("{base}.schema.{ext}")).unwrap(),
-                        &mut spec.schema,
-                    )?;
+                    // Potentially write out separate schema files for the collection
+                    // schema / write_schema / read_schema.
+                    for (schema, filename) in [
+                        (spec.schema.as_mut(), format!("{base}.schema.{ext}")),
+                        (
+                            spec.write_schema.as_mut(),
+                            format!("{base}.write-schema.{ext}"),
```
[the remark this answers is not in the document]
Comment (jgraettinger) [reply crates/flowctl/src/draft/develop.rs:108]: Agreed, will go with `{base}.write.schema.yaml`

File: crates/schemalate/src/firebolt/firebolt_schema_builder.rs (hunk 2 above)
Comment (jgraettinger) [reply crates/schemalate/src/firebolt/firebolt_schema_builder.rs]: Done

**Pull request merged**